hnn_core/parallel_backends.py

```diff
@@ -282,67 +304,51 @@ def simulate(self, net):
         os.close(pipe_stdin_w)
         os.close(pipe_stdin_r)
 
-        proc_stderr_bytes = b''
-        # capture stdout while waiting for process to complete
+        # data will be stored here; output will be printed and discarded
+        self.proc_data_bytes = b''
+
+        # create the selector instance and register all input events
+        # with self.read_stdout which will only echo to stdout
+        self.sel = selectors.DefaultSelector()
+        self.sel.register(pipe_stdout_r, selectors.EVENT_READ,
+                          self._read_stdout)
+        self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
+                          self._read_stdout)
+
+        # loop while the process is running
         while proc.poll() is None:
-            ready = select.select([pipe_stdout_r, pipe_stderr_r], [], [], 0)[0]
-            if not ready:
-                continue
-            for stream in ready:
-                if stream == pipe_stdout_r:
-                    buf = os.read(pipe_stdout_r, 1024)
-                    # write processes stdout to our stdout (fd 0)
-                    try:
-                        stdout_fd = sys.stdout.fileno()
-                    except AttributeError:
-                        # for sphinx-gallery that replaces sys.stdout
-                        stdout_fd = 0
-                    os.write(stdout_fd, buf)
-                else:
-                    proc_stderr_bytes += os.read(pipe_stderr_r, 1024)
+            # wait for an event on the selector, timeout after 1s and check
+            # that process is still running
+            events = self.sel.select(timeout=1)
+            for key, mask in events:
+                callback = key.data
+                finished = callback(key.fileobj, mask)
+                if finished:
+                    # finishied receiving printable output
+                    self.sel.unregister(pipe_stdout_r)
+                    self.sel.unregister(pipe_stderr_r)
+                    # everything received should now be handled by _read_data
+                    self.sel.register(pipe_stderr_r, selectors.EVENT_READ,
+                                      self._read_data)
```

Comment:
Before this, `_read_stdout` was called for events on stderr and stdout. Now just `_read_data` is called to handle events on stderr